File: branches/android/trunk/mobileCloud/android/2_0/test-suite/test-sync-port/src/main/java/org/openmobster/core/mobileCloud/android/module/sync/TestSlowSync.java
/******************************************************************************
 * OpenMobster                                                                *
 * Copyright 2008, OpenMobster, and individual                                *
 * contributors as indicated by the @authors tag. See the                     *
 * copyright.txt in the distribution for a full listing of                    *
 * individual contributors.                                                   *
 *                                                                            *
 * This is free software; you can redistribute it and/or modify it            *
 * under the terms of the GNU Lesser General Public License as                *
 * published by the Free Software Foundation; either version 2.1 of           *
 * the License, or (at your option) any later version.                        *
 *                                                                            *
 * This software is distributed in the hope that it will be useful,           *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU           *
 * Lesser General Public License for more details.                            *
 *                                                                            *
 * You should have received a copy of the GNU Lesser General Public           *
 * License along with this software; if not, write to the Free                *
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA         *
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.                   *
 ******************************************************************************/
package org.openmobster.core.mobileCloud.android.module.sync;


import org.openmobster.core.mobileCloud.android.module.mobileObject.MobileObject;


/**
 * 
 * @author openmobster@gmail.com
 *
 */
public final class TestSlowSync extends AbstractSyncTest 
{		
	/**
	 * 
	 */
	public void runTest()
	{
		try
		{
			//Add test case
			this.setUp("add");
			SyncService.getInstance().performSlowSync(service, service, false);			
			//These records from the server should show up on the device
			this.assertRecordPresence("unique-1", "/TestSlowSync/add");
			this.assertRecordPresence("unique-2", "/TestSlowSync/add");
			this.assertRecordPresence("unique-3", "/TestSlowSync/add");
			this.assertRecordPresence("unique-4", "/TestSlowSync/add");
			this.tearDown();
			
			//Replace test case
			this.setUp("replace");						
			SyncService.getInstance().performSlowSync(service, service, false);
			MobileObject replacedRecord = this.getRecord("unique-2");			
			//These records from the server should show up on the device
			this.assertRecordPresence("unique-1", "/TestSlowSync/replace");
			this.assertRecordPresence("unique-2", "/TestSlowSync/replace");
			this.assertEquals(replacedRecord.getValue("message"), "<tag apos='apos' quote=\"quote\" ampersand='&'>unique-2/Updated/Client</tag>", "/TestSlowSync/replace/updated");
			this.tearDown();
			
			//Delete test case
			this.setUp("delete");
			SyncService.getInstance().performSlowSync(service, service, false);							
			//These records deleted from the device should *show up* on the device
			//Should not be deleted in a SlowSync Setup
			this.assertRecordPresence("unique-1", "/TestSlowSync/delete");
			this.assertRecordPresence("unique-2", "/TestSlowSync/delete");
			this.tearDown();
			
			//Conflict test case
			this.setUp("conflict");						
			SyncService.getInstance().performSlowSync(service, service, false);
			MobileObject conflictedRecord = this.getRecord("unique-1");			
			//These records from the server should show up on the device
			this.assertRecordPresence("unique-1", "/TestSlowSync/conflict");
			this.assertRecordPresence("unique-2", "/TestSlowSync/conflict");
			this.assertEquals(conflictedRecord.getValue("message"), "<tag apos='apos' quote=\"quote\" ampersand='&'>unique-1/Updated/Client</tag>", "/TestSlowSync/conflict/unique-1");
			this.tearDown();
		}
		catch(Exception e)
		{
			e.printStackTrace(System.out);
			throw new RuntimeException(e.toString());
		}
	}	
}

File: trunk/iphone/mobilecloudlib/TestSlowSync.m
/**
 * Copyright (c) {2003,2010} {openmobster@gmail.com} {individual contributors as indicated by the @authors tag}.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

#import "TestSlowSync.h"


/**
 * 
 * @author openmobster@gmail.com
 */
@implementation TestSlowSync

+(id)withInit
{
	return [[[TestSlowSync alloc] init] autorelease];
}

-(void) runTest
{
	SyncService *sync = [SyncService getInstance];
	
	//Add testcase
	[self setUp:@"add"];
	[sync performSlowSync:channel :NO];
	[self assertRecordPresence:@"unique-1" :@"/TestSlowSync/add"];
	[self assertRecordPresence:@"unique-2" :@"/TestSlowSync/add"];
	[self assertRecordPresence:@"unique-3" :@"/TestSlowSync/add"];
	[self assertRecordPresence:@"unique-4" :@"/TestSlowSync/add"];
	[self tearDown];
	
	//Replace testcase
	[self setUp:@"replace"];
	[sync performSlowSync:channel :NO];
	MobileObject *afterUnique1 = [self getRecord:@"unique-1"];
	MobileObject *afterUnique2 = [self getRecord:@"unique-2"];
	[self assertRecordPresence:@"unique-1" :@"/TestSlowSync/replace"];
	[self assertRecordPresence:@"unique-2" :@"/TestSlowSync/replace"];
	NSString *compareMessage1 = @"<tag apos='apos' quote=\"quote\" ampersand='&'>unique-1/Message</tag>";
	NSString *compareMessage2 = @"<tag apos='apos' quote=\"quote\" ampersand='&'>unique-2/Updated/Client</tag>";
	NSString *afterUnique1Msg = [afterUnique1 getValue:@"message"];
	NSString *afterUnique2Msg = [afterUnique2 getValue:@"message"];
	NSLog(@"%@",afterUnique1Msg);
	NSLog(@"%@",afterUnique2Msg);
	[self assertTrue:[afterUnique1Msg isEqualToString:compareMessage1] :@"/TestSlowSync/replace/updated/unique-1"];
	[self assertTrue:[afterUnique2Msg isEqualToString:compareMessage2] :@"/TestSlowSync/replace/updated/unique-2"];
	[self tearDown];
	
	//Delete testcase
	[self setUp:@"delete"];
	[sync performSlowSync:channel :NO];
	[self assertRecordPresence:@"unique-1" :@"/TestSlowSync/delete"];
	[self assertRecordPresence:@"unique-2" :@"/TestSlowSync/delete"];
	[self tearDown];
	
	//Conflict testcase
	[self setUp:@"conflict"];
	[sync performSlowSync:channel :NO];
	afterUnique1 = [self getRecord:@"unique-1"];
	afterUnique2 = [self getRecord:@"unique-2"];
	[self assertRecordPresence:@"unique-1" :@"/TestSlowSync/conflict"];
	[self assertRecordPresence:@"unique-2" :@"/TestSlowSync/conflict"];
	compareMessage1 = @"<tag apos='apos' quote=\"quote\" ampersand='&'>unique-1/Updated/Client</tag>";
	compareMessage2 = @"<tag apos='apos' quote=\"quote\" ampersand='&'>unique-2/Message</tag>";
	afterUnique1Msg = [afterUnique1 getValue:@"message"];
	afterUnique2Msg = [afterUnique2 getValue:@"message"];
	NSLog(@"%@",afterUnique1Msg);
	NSLog(@"%@",afterUnique2Msg);
	[self assertTrue:[afterUnique1Msg isEqualToString:compareMessage1] :@"/TestSlowSync/conflict/updated/unique-1"];
	[self assertTrue:[afterUnique2Msg isEqualToString:compareMessage2] :@"/TestSlowSync/conflict/updated/unique-2"];
	[self tearDown];
}

-(NSString *) getInfo
{
	return [[self class] description];
}
@end
